File: app/csharp/AccountingSystem.Infrastructure/Persistence/Repositories/JournalEntryReadModelRepository.cs
namespace AccountingSystem.Infrastructure.Persistence.Repositories;

using AccountingSystem.Application.Ports.Out;
using AccountingSystem.Domain.Entities;
using Dapper;
using Npgsql;

/// <summary>
/// 仕訳 Read Model リポジトリ実装
/// CQRS パターンの Query Side 用
/// </summary>
public class JournalEntryReadModelRepository : IJournalEntryReadModelRepository
{
    private readonly string _connectionString;

    public JournalEntryReadModelRepository(string connectionString)
    {
        _connectionString = connectionString;
    }

    public async Task InsertJournalEntryAsync(
        string id,
        DateOnly entryDate,
        string description,
        string status,
        bool deleted,
        DateTime createdAt,
        DateTime updatedAt,
        string? approvedBy,
        string? approvalComment)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var sql = @"
            INSERT INTO ""仕訳ReadModel"" (
                ""仕訳ID"", ""仕訳日"", ""摘要"", ""ステータス"", ""削除済み"",
                ""作成日時"", ""更新日時"", ""承認者"", ""承認コメント""
            ) VALUES (
                @Id, @EntryDate, @Description, @Status, @Deleted,
                @CreatedAt, @UpdatedAt, @ApprovedBy, @ApprovalComment
            )
            ";

        await connection.ExecuteAsync(sql, new
        {
            Id = id,
            EntryDate = entryDate,
            Description = description,
            Status = status,
            Deleted = deleted,
            CreatedAt = createdAt,
            UpdatedAt = updatedAt,
            ApprovedBy = approvedBy,
            ApprovalComment = approvalComment
        });
    }

    public async Task InsertJournalEntryLineAsync(
        string journalEntryId,
        string accountCode,
        string debitCredit,
        decimal amount)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var sql = @"
            INSERT INTO ""仕訳明細ReadModel"" (
                ""仕訳ID"", ""勘定科目コード"", ""貸借区分"", ""金額""
            ) VALUES (
                @JournalEntryId, @AccountCode, @DebitCredit, @Amount
            )
            ";

        await connection.ExecuteAsync(sql, new
        {
            JournalEntryId = journalEntryId,
            AccountCode = accountCode,
            DebitCredit = debitCredit,
            Amount = amount
        });
    }

    public async Task UpdateJournalEntryStatusAsync(
        string id,
        string status,
        DateTime updatedAt,
        string approvedBy,
        string approvalComment)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var sql = @"
            UPDATE ""仕訳ReadModel""
            SET ""ステータス"" = @Status,
                ""更新日時"" = @UpdatedAt,
                ""承認者"" = @ApprovedBy,
                ""承認コメント"" = @ApprovalComment
            WHERE ""仕訳ID"" = @Id
            ";

        await connection.ExecuteAsync(sql, new
        {
            Id = id,
            Status = status,
            UpdatedAt = updatedAt,
            ApprovedBy = approvedBy,
            ApprovalComment = approvalComment
        });
    }

    public async Task MarkAsDeletedAsync(string id, DateTime updatedAt)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var sql = @"
            UPDATE ""仕訳ReadModel""
            SET ""削除済み"" = true, ""更新日時"" = @UpdatedAt
            WHERE ""仕訳ID"" = @Id
            ";

        await connection.ExecuteAsync(sql, new { Id = id, UpdatedAt = updatedAt });
    }

    public async Task<JournalEntryReadModel?> SelectByIdAsync(string id)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var sql = @"
            SELECT ""仕訳ID"" as Id,
                   ""仕訳日"" as EntryDate,
                   ""摘要"" as Description,
                   ""ステータス"" as Status,
                   ""削除済み"" as Deleted,
                   ""作成日時"" as CreatedAt,
                   ""更新日時"" as UpdatedAt,
                   ""承認者"" as ApprovedBy,
                   ""承認コメント"" as ApprovalComment
            FROM ""仕訳ReadModel""
            WHERE ""仕訳ID"" = @Id
            ";

        return await connection.QuerySingleOrDefaultAsync<JournalEntryReadModel>(sql, new { Id = id });
    }

    public async Task<IReadOnlyList<JournalEntryReadModel>> SelectByDateRangeAsync(
        DateOnly startDate,
        DateOnly endDate)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var sql = @"
            SELECT ""仕訳ID"" as Id,
                   ""仕訳日"" as EntryDate,
                   ""摘要"" as Description,
                   ""ステータス"" as Status,
                   ""削除済み"" as Deleted,
                   ""作成日時"" as CreatedAt,
                   ""更新日時"" as UpdatedAt,
                   ""承認者"" as ApprovedBy,
                   ""承認コメント"" as ApprovalComment
            FROM ""仕訳ReadModel""
            WHERE ""仕訳日"" >= @StartDate
              AND ""仕訳日"" <= @EndDate
              AND ""削除済み"" = false
            ORDER BY ""仕訳日"", ""仕訳ID""
            ";

        var results = await connection.QueryAsync<JournalEntryReadModel>(
            sql, new { StartDate = startDate, EndDate = endDate });
        return results.ToList();
    }

    public async Task<IReadOnlyList<JournalEntryLineReadModel>> SelectLinesByJournalEntryIdAsync(
        string journalEntryId)
    {
        await using var connection = new NpgsqlConnection(_connectionString);
        await connection.OpenAsync();

        var sql = @"
            SELECT ""明細ID"" as Id,
                   ""仕訳ID"" as JournalEntryId,
                   ""勘定科目コード"" as AccountCode,
                   ""貸借区分"" as DebitCredit,
                   ""金額"" as Amount
            FROM ""仕訳明細ReadModel""
            WHERE ""仕訳ID"" = @JournalEntryId
            ORDER BY ""明細ID""
            ";

        var results = await connection.QueryAsync<JournalEntryLineReadModel>(
            sql, new { JournalEntryId = journalEntryId });
        return results.ToList();
    }
}

File: app/fsharp/AccountingSystem.Infrastructure/Persistence/Repositories/JournalEntryReadModelRepository.fs
namespace AccountingSystem.Infrastructure.Persistence.Repositories

open System
open System.Threading.Tasks
open Npgsql
open Dapper
open AccountingSystem.Application.Port.Out
open AccountingSystem.Infrastructure.Persistence.DAO

/// <summary>
/// 仕訳 Read Model リポジトリ実装（Dapper）
/// </summary>
type JournalEntryReadModelRepository(connection: NpgsqlConnection) =

    /// エンティティをドメインモデルに変換
    let toReadModel (entity: JournalEntryReadModelEntity) : JournalEntryReadModel =
        {
            Id = entity.id
            EntryDate = entity.entry_date
            Description = entity.description
            Status = entity.status
            Deleted = entity.deleted
            CreatedAt = entity.created_at
            UpdatedAt = entity.updated_at
            ApprovedBy = if isNull entity.approved_by then None else Some entity.approved_by
            ApprovalComment = if isNull entity.approval_comment then None else Some entity.approval_comment
        }

    /// 明細エンティティをドメインモデルに変換
    let toLineReadModel (entity: JournalEntryLineReadModelEntity) : JournalEntryLineReadModel =
        {
            Id = entity.id
            JournalEntryId = entity.journal_entry_id
            AccountCode = entity.account_code
            DebitCredit = entity.debit_credit
            Amount = entity.amount
        }

    interface IJournalEntryReadModelRepository with
        member _.InsertJournalEntryAsync
            (id: string)
            (entryDate: DateTime)
            (description: string)
            (status: string)
            (deleted: bool)
            (createdAt: DateTime)
            (updatedAt: DateTime)
            (approvedBy: string option)
            (approvalComment: string option)
            : Task<unit> =
            task {
                let sql = """
                    INSERT INTO journal_entry_read_model (
                        id, entry_date, description, status, deleted,
                        created_at, updated_at, approved_by, approval_comment
                    ) VALUES (
                        @Id, @EntryDate, @Description, @Status, @Deleted,
                        @CreatedAt, @UpdatedAt, @ApprovedBy, @ApprovalComment
                    )
                """

                let! _ = connection.ExecuteAsync(sql, {|
                    Id = id
                    EntryDate = entryDate
                    Description = description
                    Status = status
                    Deleted = deleted
                    CreatedAt = createdAt
                    UpdatedAt = updatedAt
                    ApprovedBy = approvedBy |> Option.toObj
                    ApprovalComment = approvalComment |> Option.toObj
                |})

                return ()
            }

        member _.InsertJournalEntryLineAsync
            (journalEntryId: string)
            (accountCode: string)
            (debitCredit: string)
            (amount: decimal)
            : Task<unit> =
            task {
                let sql = """
                    INSERT INTO journal_entry_line_read_model (
                        journal_entry_id, account_code, debit_credit, amount
                    ) VALUES (
                        @JournalEntryId, @AccountCode, @DebitCredit, @Amount
                    )
                """

                let! _ = connection.ExecuteAsync(sql, {|
                    JournalEntryId = journalEntryId
                    AccountCode = accountCode
                    DebitCredit = debitCredit
                    Amount = amount
                |})

                return ()
            }

        member _.UpdateJournalEntryStatusAsync
            (id: string)
            (status: string)
            (updatedAt: DateTime)
            (approvedBy: string)
            (approvalComment: string)
            : Task<unit> =
            task {
                let sql = """
                    UPDATE journal_entry_read_model
                    SET status = @Status,
                        updated_at = @UpdatedAt,
                        approved_by = @ApprovedBy,
                        approval_comment = @ApprovalComment
                    WHERE id = @Id
                """

                let! _ = connection.ExecuteAsync(sql, {|
                    Id = id
                    Status = status
                    UpdatedAt = updatedAt
                    ApprovedBy = approvedBy
                    ApprovalComment = approvalComment
                |})

                return ()
            }

        member _.MarkAsDeletedAsync (id: string) (updatedAt: DateTime) : Task<unit> =
            task {
                let sql = """
                    UPDATE journal_entry_read_model
                    SET deleted = true, updated_at = @UpdatedAt
                    WHERE id = @Id
                """

                let! _ = connection.ExecuteAsync(sql, {| Id = id; UpdatedAt = updatedAt |})

                return ()
            }

        member _.SelectByIdAsync (id: string) : Task<JournalEntryReadModel option> =
            task {
                let sql = """
                    SELECT id, entry_date, description, status, deleted,
                           created_at, updated_at, approved_by, approval_comment
                    FROM journal_entry_read_model
                    WHERE id = @Id
                """

                let! entity = connection.QuerySingleOrDefaultAsync<JournalEntryReadModelEntity>(sql, {| Id = id |})

                return if isNull (box entity) then None else Some (toReadModel entity)
            }

        member _.SelectByDateRangeAsync (startDate: DateTime) (endDate: DateTime) : Task<JournalEntryReadModel list> =
            task {
                let sql = """
                    SELECT id, entry_date, description, status, deleted,
                           created_at, updated_at, approved_by, approval_comment
                    FROM journal_entry_read_model
                    WHERE entry_date BETWEEN @StartDate AND @EndDate
                      AND deleted = false
                    ORDER BY entry_date DESC
                """

                let! entities = connection.QueryAsync<JournalEntryReadModelEntity>(sql, {|
                    StartDate = startDate
                    EndDate = endDate
                |})

                return entities |> Seq.map toReadModel |> Seq.toList
            }

        member _.SelectLinesByJournalEntryIdAsync (journalEntryId: string) : Task<JournalEntryLineReadModel list> =
            task {
                let sql = """
                    SELECT id, journal_entry_id, account_code, debit_credit, amount
                    FROM journal_entry_line_read_model
                    WHERE journal_entry_id = @JournalEntryId
                """

                let! entities = connection.QueryAsync<JournalEntryLineReadModelEntity>(sql, {|
                    JournalEntryId = journalEntryId
                |})

                return entities |> Seq.map toLineReadModel |> Seq.toList
            }

        member _.SelectAllAsync () : Task<JournalEntryReadModel list> =
            task {
                let sql = """
                    SELECT id, entry_date, description, status, deleted,
                           created_at, updated_at, approved_by, approval_comment
                    FROM journal_entry_read_model
                    WHERE deleted = false
                    ORDER BY entry_date DESC
                """

                let! entities = connection.QueryAsync<JournalEntryReadModelEntity>(sql)

                return entities |> Seq.map toReadModel |> Seq.toList
            }
